packages/types/src/data/abstract-query.ts

```diff
@@ -0,0 +1,240 @@
+
+	/** Name of the collection within the datastore */
+	collection: string;
+
+	/** All fields to select in the query */
+	fieldNodes: AbstractQueryFieldNode[];
+
+	// modifiers?: {
+	// 	limit?: AbstractQueryNodeLimit;
+	// 	filter?: AbstractQueryFilterNode[];
+	// };
+}
+
+type AbstractQueryNodeType = 'primitive' | 'fn' | 'm2o' | 'o2m' | 'a2o' | 'o2a';
+
+interface AbstractQueryNode {
+	/** the type of the node */
+	type: AbstractQueryNodeType;
+}
+
+/** A group of all possible field types */
+export type AbstractQueryFieldNode =
+	| AbstractQueryFieldNodePrimitive
+	| AbstractQueryFieldNodeFn
+	| AbstractQueryFieldNodeRelated;
+
+/** Generic primitive value read from the datastore field */
+export interface AbstractQueryFieldNodePrimitive extends AbstractQueryNode {
+	type: 'primitive';
+
+	/** the name of the attribute */
+	field: string;
+}
+
+export type AbstractQueryFn = 'year' | 'month' | 'week' | 'day' | 'weekday' | 'hour' | 'minute' | 'second';
+
+export interface AbstractQueryFieldNodeFn extends AbstractQueryNode {
+	type: 'fn';
+
+	fn: AbstractQueryFn;
+
+	targetNode: AbstractQueryFieldNodePrimitive | AbstractQueryFieldNodeFn;
+
+	args?: (string | number | boolean)[];
+}
+
+export interface AbstractQueryFieldNodeRelatedBase {
+	fieldNodes: AbstractQueryFieldNode[];
+
+	modifiers?: AbstractQueryModifiers;
+}
+
+export type AbstractQueryFieldNodeRelated =
+	| AbstractQueryFieldNodeRelatedManyToOne
+	| AbstractQueryFieldNodeRelatedOneToMany
+	| AbstractQueryFieldNodeRelatedAnyToOne
+	| AbstractQueryFieldNodeRelatedOneToAny;
+
+interface AbstractQueryFieldNodeRelatedJoinMany {
+	current: {
+		fields: string[];
+	};
+
+	external: {
+		datastore?: string;
+		collection: string;
+		fields: string[];
+	};
+}
+
+interface AbstractQueryFieldNodeRelatedJoinAny {
+	current: {
+		collectionField: string;
+		fields: string[];
+	};
+
+	external: {
+		datastore?: string;
+		fields: string[];
+	};
+}
+
+export interface AbstractQueryFieldNodeRelatedManyToOne extends AbstractQueryNode, AbstractQueryFieldNodeRelatedBase {
+	type: 'm2o';
+
+	join: AbstractQueryFieldNodeRelatedJoinMany;
+}
+
+export interface AbstractQueryFieldNodeRelatedOneToMany extends AbstractQueryNode, AbstractQueryFieldNodeRelatedBase {
+	type: 'o2m';
+
+	join: AbstractQueryFieldNodeRelatedJoinMany;
+}
+
+export interface AbstractQueryFieldNodeRelatedAnyToOne extends AbstractQueryNode, AbstractQueryFieldNodeRelatedBase {
+	type: 'a2o';
+
+	join: AbstractQueryFieldNodeRelatedJoinAny;
+}
+
+export interface AbstractQueryFieldNodeRelatedOneToAny extends AbstractQueryNode, AbstractQueryFieldNodeRelatedBase {
+	type: 'o2a';
+
+	join: AbstractQueryFieldNodeRelatedJoinAny;
+}
+
+// =================================================================================================
+
+/**
+ * Optional attributes to customize the query results
+ * @typeParam sort - Specifies the order of the results
+ * @typeParam limit - Specifies the maximum amount of returning results
+ */
```

Comment:
based on the interface, should we include `offset` and `filter` in this tsdoc as well? (and possibly order it based on their order in the interface)

Reply:
I now removed all of those since they are fairly obvious what they mean